src/StackExchange.Redis/ServerEndPoint.cs

```diff
@@ -77,6 +76,32 @@ public ServerEndPoint(ConnectionMultiplexer multiplexer, EndPoint endpoint)
 
         public bool IsConnecting => interactive?.IsConnecting == true;
 
+        private readonly List<TaskCompletionSource<bool>> _pendingConnectionMonitors = new List<TaskCompletionSource<bool>>();
+
+        /// <summary>
+        /// Awaitable state seeing if this endpoint is connected.
+        /// </summary>
+        public Task<bool> OnConnectedAsync(LogProxy log = null, bool sendTracerIfConnected = false)
+        {
+            if (!IsConnected)
```

Comment:
Looks like we once again could get in trouble with a potential race:
-observe IsConnect = false
-head down Setup the task completion sources part of the if statement
-in other thread, Connection completes, acquires the lock, sees no handlers, and release the lock
-acquire the lock actually finish adding task completion sources, and release the lock

Reply:
Agreed it can happen, though the chances are slim the the effect there would be waiting the full connection time allotted and complete successfully. We could add another `IsConnected` check and `.TrySetResult(true);` immediately here as well - thoughts?

Reply:
How's this for an idea:
What if we replace the PendingConnectionMonitors list with a single TaskCompletionSource property WhenConnectedAwaitable.

A new TaskCompletionSource can be atomically swapped in every time we try to recreate the connection. It can be null, or remain at the previous value, until then, and once the connection attempt completes or fails, it becomes a completed task.

Anyone attempting to wait while no connection attempt yet in progress can get nullrefs (OK maybe we should improve on that? Give them a more specific exception? :-/ )

Anyone awaiting on it should either eventually get successfully completed task (connected!) or an exception (connection failed!)

Reply:
My worry there is we'd be creating a task creation/swap race and potentially uncompleted tasks forever or an await in the multiplexer attached to the old source that went out of scope during a reconnect on the underlying bridge, ya know? That's why I figured a list was safest, to leave nothing unresolved.